sequence_transform.c

```diff
@@ -288,29 +175,24 @@ void handle_repeat_key()
     uprintf("repeat keycode: 0x%04X\n", keycode);
 #endif
     if (keycode && !(keycode & SPECIAL_KEY_TRIECODE_0)) {
-        dequeue_keycodes(1);
-        enqueue_keycode(keycode);
+        key_buffer_dequeue(&key_buffer, 1);
+        key_buffer_enqueue(&key_buffer, keycode);
         send_key(keycode);
     }
 }
 //////////////////////////////////////////////////////////////////////////////////////////
-void handle_result(trie_t *trie, trie_search_result_t *res)
+void handle_result(trie_t *trie, trie_payload_t *res)
 {
 #ifdef SEQUENCE_TRANSFORM_LOG_GENERAL
-    uprintf("search res: offset: %d, len: %d, bspaces: %d, func: %d\n",
-            res->completion_offset, res->complete_len, res->num_backspaces, res->func_code);
+    uprintf("completion search res: index: %d, len: %d, bspaces: %d, func: %d\n",
+            res->completion_index, res->completion_len, res->num_backspaces, res->func_code);
 #endif
-    // bounds check completion data
-    if (res->completion_offset + res->complete_len > trie->completions_size) {
-        uprintf("ERROR: trying to read past end of completion data buffer! offset: %d, len: %d\n",
-                res->completion_offset, res->complete_len);
-        return;
-    }
     // Send backspaces
     multi_tap(KC_BSPC, res->num_backspaces);
     // Send completion string
-    bool ends_with_wordbreak = (res->complete_len > 0 && CDATA(res->completion_offset + res->complete_len - 1) == ' ');
-    for (uint16_t i = res->completion_offset, end = i+res->complete_len; i < end; ++i) {
+    const uint16_t completion_end = res->completion_index + res->completion_len;
+    bool ends_with_wordbreak = (res->completion_len > 0 && CDATA(completion_end - 1) == ' ');
+    for (uint16_t i = res->completion_index; i < completion_end; ++i) {
```

Comment:
This for loop no longer performs a sanity check on the integrity of the completions array, and will read past the end of the array if the `completion_len` is too long.

Since these values are known at compile time, we could run an integrity check at compile time. The most likely reason to lose integrity would be a user using a `sequence_transform_data.h` file generated by an incompatible version of the generator that mapped `complete_len` differently than the qmk code expects. So, it is important that we don't rely on the python script to perform the sanity check. Sidenote: mismatching the generator and qmk library is a huge potential source for user error in this library, so I'm going to add a version check.